php/libraries/SubprojectController.class.inc

```diff
@@ -0,0 +1,252 @@
+ */
+
+namespace LORIS;
+
+/**
+ * Base class that contains method to access the subproject of projects
+ *
+ * @category Controller
+ * @package  Subprojects
+ * @author   Rida Abou-Haidar <rida.abou-haidar@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class SubprojectController
+{
+    /**
+     * Array holding existing instances
+     *
+     * @var array
+     */
+    private static $_instances = [];
+
+    /**
+     * The database connection to use
+     *
+     * @var \Database
+     */
+    protected $database;
+
+    /**
+     * The constructor
+     *
+     * @param \Database $database The database connection to use
+     */
+    public function __construct(\Database $database)
+    {
+        if (!$database->isConnected()) {
+            throw new \RuntimeException(
+                "SubprojectController - Database not connected"
+            );
+        }
+        $this->database = $database;
+    }
+
+    /**
+     * Retruns a list of configured Subprojects
+     *
+     * @return array The list of subprojects
+     */
+    public function getAllSubprojects(): array
+    {
+        $subprojectRows = $this->database->pselect(
+            "SELECT
+                SubprojectID,
+                title,
+                useEDC,
+                WindowDifference,
+                RecruitmentTarget
+            FROM subproject
+            ",
+            []
+        );
+
+        $subprojects = [];
+        foreach ($subprojectRows as $row) {
+            $subproject    = new \LORIS\Subproject(
+                intval($row['SubprojectID']),
+                $row['title'],
+                $row['useEDC'] === '1',
+                $row['WindowDifference'],
+                isset($row['RecruitmentTarget'])
+                    ? intval($row['RecruitmentTarget']) : null
+            );
+            $subprojects[] = $subproject;
+        }
+
+        return $subprojects;
+    }
+
+    /**
+     * This returns a single subproject object for the given subproject
+     * title and will only ever instantiate
+     * one object in the event that it's called multiple times.
+     *
+     * @param string $subprojectTitle The title of the subproject
+     *
+     * @return \LORIS\Subproject the Subproject object
+     */
+    public function getSubprojectByTitle($subprojectTitle) : \LORIS\Subproject
+    {
+        if (!isset(self::$_instances[$subprojectTitle])) {
+            $subprojectData = $this->database->pSelectRow(
+                "SELECT
+                    SubprojectID,
+                    title,
+                    useEDC,
+                    WindowDifference,
+                    RecruitmentTarget
+                FROM subproject
+                WHERE title = :title",
+                ['title' => $subprojectTitle]
+            );
+
+            if (empty($subprojectData)) {
+                throw new \LorisException(
```

Comment:
We have a NotFound exception

Reply:
I know but I thought not found was for files not found on system or something... not database entities (I actually had it as notfound and changed it)

Reply:
NotFound exceptions tend to get converted to 404 errors by middleware.

That should probably be the criteria you use to determine if it makes sense to use it.

Reply:
@driusan @xlecours i'm not sure that answers my question. if you query for a candidate that doesnt exist... is that a 404?? i dont think so no ?

Reply:
will change to notFound